[CaptureTracking][NFC] Clarify usage expectations in PointerMayBeCaptured comments

Camsyn opened the pull request:
Fixes issue #132739.

CaptureAnalysis only considers captures through the def-use chain of the provided pointer, explicitly excluding captures of underlying values or implicit captures like those involving external globals. 

The previous comment for `PointerMayBeCaptured` did not clearly state this limitation, leading to its incorrect usage in files such as ThreadSanitizer.cpp and SanitizerMetadata.cpp. 

This PR addresses this by refining the comments for the relevant APIs within `PointerMayBeCaptured` to explicitly document this behavior.

Camsyn committed:
[CaptureTracking] Supports analysis for dervied pointers

Fixes issue #132739.

This fixes a missed capture detection scenario where only the base
pointer is captured.

Previously, PointerMayBeCaptured only performed forward def-use
analysis on the given pointer, which worked for base pointers
(Arguments/Allocas/Calls) but failed to detect captures of derived
pointers when their base pointers were captured.

The fix: For the input pointer, first trace back to its base pointer
using `getUnderlyingObject`, then perform capture analysis on the base.
This ensures proper handling of derived pointers with bases captured.

Performance considerations:
- Most existing callers already pass base pointers (the common case), so
the added backtracing has negligible overhead
- The few callers (ThreadSanitizer.cpp/SanitizerBinaryMetadata.cpp)
passing arbitrary pointers are sanitizer-related components where
compilation time is less critical compared to runtime overhead

This addresses false negatives in capture detection while maintaining
 reasonable compilation efficiency.

antoniofrighetto commented on llvm/unittests/Analysis/CaptureTrackingTest.cpp:
```diff
@@ -132,3 +132,31 @@ TEST(CaptureTracking, MultipleUsesInSameInstruction) {
+    declare void @bar(ptr)
+
+    define void @test() {
+      %stkobj = alloca [2 x i32]
```
```suggestion
      %stack_obj = alloca [2 x i32]
```

antoniofrighetto commented on llvm/unittests/Analysis/CaptureTrackingTest.cpp:
```diff
@@ -132,3 +132,31 @@ TEST(CaptureTracking, MultipleUsesInSameInstruction) {
+  Function *F = M->getFunction("test");
+  BasicBlock *BB = &F->getEntryBlock();
+  Instruction *StackObj = &*BB->begin();
+  Instruction *DerviedPtr = StackObj->getNextNode();
```
```suggestion
  Instruction *DerivedPtr = StackObj->getNextNode();
```

antoniofrighetto reviewed (commented):
Mind sharing the compile-time impact for this change?

Camsyn committed:
Tackle some format issues adhering the suggestions of the reviewer

Camsyn commented:
> Mind sharing the compile-time impact for this change?

Thanks @dtcxzyw for his performance testing. In the [llvm-opt-benchmark](https://github.com/dtcxzyw/llvm-opt-benchmark) he built, this modification brings **~0.01%** compilation overhead.

I'm trying to rebuild clang/LLVM in Release mode and compile tests of llvm-test-suite locally. If there are any results, I'll sync them.

nikic reviewed (changes requested):
We definitely should not be doing this in CaptureTracking -- passing a sensible value is the responsibility of the caller. In most contexts, performing capture checks only makes sense on "identified function-local objects".

nikic commented:
Sure, expanding the comment on the function is a good idea.

Camsyn committed:
Revert "Tackle some format issues adhering the suggestions of the reviewer"

This reverts commit cd64fa32a925de801c77864e71f49796f5193c87.

Camsyn committed:
Revert "[CaptureTracking] Supports analysis for dervied pointers"

This reverts commit 87c0937b9cc29c312a379a72c07aea84a8dcbdf0.

Camsyn committed:
[CaptureTracking][NFC] Clarify usage expectations in PointerMayBeCaptured comments

The comments for PointerMayBeCaptured and related APIs have been
updated to explicitly state that these functions assume the input is an
function-local object. It is the caller's responsibility to ensure this
precondition is met. This clarification aligns with recent discussions
and feedback from code reviews, aiming to prevent misuse and potential
analysis errors.

Camsyn commented:
Since I didn’t see the comment update in the latest commit yet, I’ve gone ahead and included it in this PR — hope that’s okay.

nikic commented on llvm/include/llvm/Analysis/CaptureTracking.h:
```diff
@@ -41,13 +41,17 @@ namespace llvm {
   /// one value before giving up due too "too many uses". If MaxUsesToExplore
   /// is zero, a default value is assumed.
+  /// This assumes the pointer is to a function-local object. The caller is
+  /// responsible for ensuring this.
```
This depends on the use case. In most cases you'd want to pass a function-local object, but e.g. to infer captures() attribute you'd just work with arbitrary function arguments.

I think I'd add something like this instead:

> This function only considers captures of the passed value via its use-def chain, without considering captures of values it may be based on, or implicit captures such as for external globals.

Camsyn replied:
> This depends on the use case. In most cases you'd want to pass a function-local object, but e.g. to infer captures() attribute you'd just work with arbitrary function arguments.
> 
> I think I'd add something like this instead:
> 
> > This function only considers captures of the passed value via its use-def chain, without considering captures of values it may be based on, or implicit captures such as for external globals.

Good point, and thank you for the improved phrasing! Your suggested comment is indeed more appropriate. Just a minor correction: it should be 'def-use chain' rather than 'use-def chain' because our CaptureAnalysis works by forward tracking using `getUser`.

Camsyn committed:
Refine the comment aadhering to the reviewer

nikic reviewed (approved):
LGTM

Camsyn committed:
Merge branch 'main' into fix-capture
